src/persist-client/src/internal/compact.rs

```diff
@@ -469,34 +469,72 @@ where
+        for (part_desc, parts) in &mut runs {
             if let Some(part) = parts.next() {
-                let start = Instant::now();
-                let mut part = fetch_batch_part(
-                    shard_id,
-                    blob.as_ref(),
-                    &metrics,
-                    &metrics.read.compaction,
-                    &part.key,
-                    part_desc,
-                )
-                .await?;
-                timings.part_fetching += start.elapsed();
-                let start = Instant::now();
-                while let Some((k, v, mut t, d)) = part.next() {
-                    t.advance_by(desc.since().borrow());
-                    let d = D::decode(d);
-                    let k = k.to_vec();
-                    let v = v.to_vec();
-                    // default heap ordering is descending
-                    sorted_updates.push(Reverse((((k, v), t, d), index)));
-                    remaining_updates_by_run[index] += 1;
+                let shard_id = shard_id.clone();
+                let blob = Arc::clone(&blob);
+                let metrics = Arc::clone(&metrics);
+                let compaction_metrics = metrics.read.compaction.clone();
+                let key = part.key.clone();
+                let part_desc = (**part_desc).clone();
+
+                pending_initial_parts.push_back(mz_ore::task::spawn(
+                    || "compaction::initial_fetch_batch_part",
+                    async move {
+                        fetch_batch_part(
+                            &shard_id,
+                            blob.as_ref(),
+                            &metrics,
+                            &compaction_metrics,
+                            &key,
+                            &part_desc,
+                        )
+                        .await
+                    },
+                ));
+
+                if pending_initial_parts.len() > cfg.compaction_reads_max_outstanding_parts {
+                    let start = Instant::now();
+                    let part = pending_initial_parts
+                        .pop_front()
+                        .expect("pop failed when len was just > some usize")
+                        .await??;
+                    timings.part_fetching += start.elapsed();
+                    fetched_initial_parts.push(part);
                 }
-                timings.heap_population += start.elapsed();
             }
         }
 
+        for part in pending_initial_parts {
+            let start = Instant::now();
+            fetched_initial_parts.push(part.await??);
+            timings.part_fetching += start.elapsed();
+        }
+
+        // serially add each fetched part into our heap. this will momentarily 2x
+        // the memory required to store a given part, so this is done one at a time.
+        let start = Instant::now();
+        for (index, mut part) in fetched_initial_parts.into_iter().enumerate() {
```

Comment:
for reviewers: the absolutely imperative detail here is that `fetched_initial_parts` must match the original ordering of `runs`, as an array index is used to identify which run to refill the heap from when this part's updates are exhausted